CoreScriptsRoot/Modules/NewChat.lua

```diff
@@ -0,0 +1,233 @@
+			end
+		end
+
+		function moduleApiTable:SetVisible(visible)
+			ChatWindowState.Visible = visible
+			local event = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "SetVisible", "BindableEvent")
+			if (event) then
+				event:Fire(ChatWindowState.Visible)
+			else
+				moduleApiTable.VisibilityStateChanged:fire(ChatWindowState.Visible)
+			end
+		end
+
+		function moduleApiTable:FocusChatBar()
+			local event = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "FocusChatBar", "BindableEvent")
+			if (event) then
+				event:Fire()
+			end
+		end
+
+		function moduleApiTable:GetVisibility()
+			local func = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "GetVisibility", "BindableFunction")
+			if (func) then
+				return func:Invoke()
+			end
+			return ChatWindowState.Visible
+		end
+
+		function moduleApiTable:GetMessageCount()
+			local func = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "GetMessageCount", "BindableFunction")
+			if (func) then
+				return func:Invoke()
+			end
+			return ChatWindowState.MessageCount
+		end
+
+		function moduleApiTable:TopbarEnabledChanged(enabled)
+			ChatWindowState.TopbarEnabled = enabled
+			local event = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "TopbarEnabledChanged", "BindableEvent")
+			if (event) then
+				event:Fire(ChatWindowState.TopbarEnabled)
+			end
+		end
+
+		function moduleApiTable:IsFocused(useWasFocused)
+			local func = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "IsFocused", "BindableFunction")
+			if (func) then
+				return func:Invoke(useWasFocused)
+			end
+			return false
+		end
+		
+		moduleApiTable.ChatBarFocusChanged = Util.Signal()
+		moduleApiTable.VisibilityStateChanged = Util.Signal()
+		moduleApiTable.MessagesChanged = Util.Signal()
+
+		StarterGui.CoreGuiChangedSignal:connect(function(coreGuiType, enabled)
+			if (coreGuiType == Enum.CoreGuiType.All or coreGuiType == Enum.CoreGuiType.Chat) then
+				ChatWindowState.CoreGuiEnabled = enabled
+
+				local event = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "CoreGuiEnabled", "BindableEvent")
+				if (event) then
+					event:Fire(ChatWindowState.CoreGuiEnabled)
+				end
+			end
+		end)
+
+		GuiService:AddSpecialKey(Enum.SpecialKey.ChatHotkey)
+		GuiService.SpecialKeyPressed:connect(function(key, modifiers)
+			local event = FindIndexInCollectionWithType(communicationsConnections.ChatWindow, "SpecialKeyPressed", "BindableEvent")
+			if (event) then
+				event:Fire(key, modifiers)
+			end
+		end)
+
+		StarterGui:RegisterSetCore("ChatMakeSystemMessage", function(data)
+			local event = FindIndexInCollectionWithType(communicationsConnections.SetCore, "ChatMakeSystemMessage", "BindableEvent")
+			if (event) then
+				event:Fire(data)
+			else
+				table.insert(MakeSystemMessageCache, data)
+			end
+		end)
+		
+		StarterGui:RegisterGetCore("ChatWindowPosition", function(data)
```

Comment:
All these functions seem very copy pasted, we should refactor them into the one function that we call in these functions.